llvm/lib/Target/Sparc/SparcISelLowering.cpp

```diff
@@ -2811,12 +2818,17 @@ static SDValue LowerDYNAMIC_STACKALLOC(SDValue Op, SelectionDAG &DAG,
     regSpillArea = 96;
   }
 
-  unsigned SPReg = SP::O6;
-  SDValue SP = DAG.getCopyFromReg(Chain, dl, SPReg, VT);
-  SDValue NewSP = DAG.getNode(ISD::SUB, dl, VT, SP, Size); // Value
-  Chain = DAG.getCopyToReg(SP.getValue(1), dl, SPReg, NewSP);    // Output chain
+  AlignedPtr = DAG.getNode(ISD::SUB, dl, VT, AlignedPtr, Size);
```

Comment:
The subtraction should be done before aligning to avoid excessive stack usage.
Take for example
`old sp = 40, size = 8, align = 16`
The current formula gives:
`new sp = (40 - 16) & -16 - 8 = 8`
That is, it allocates 24 bytes instead of just 8 (which would be sufficient because `40 - 8 = 32` is 16-byte aligned).

Reply:
It should be `(ptr - size) & -align`.